### ISSUE TITLE
CPU and hard drive utilization limit per folder

### ISSUE THREAD

**Issue by Avamander** [syncthing/syncthing#3212]
I am running Syncthing on two fairly low-powered devices (hash performance 80MB/s and 36MB/s) and I've got a folder with 169403 items (9.15GiB), every time syncthing starts "scanning" the folder it makes both machines almost unusable.

This makes me unable to run syncthing in the background, other folders are smaller, it is maximum a few minutes it takes to sync other folderd but that folder it takes about 5-20 minutes (as mentioned earlier, making the system almost unusable). 

So my suggestion is to add a parameter "Sync maximum speed if idle (and not on battery), sync using x% of disk utilization and y% of CPU when not idle", idle meaning the PC is locked/no mouse movement a while.

**Comment by AudriusButkevicius** [comment]
You can probably script something up which just starts and stops syncthing based on whatever conditions you want externally.

Given the vast range of platforms we support, I think it's unlikely we'll ever support this.

**Comment by calmh** [comment]
I think we might at some point implement throttling of the scanning process. I could see that being in terms of files/s (thus potentially sleeping between steps in recursive walk function) and MB/s (when hashing). Tuning that to desired values would be an exercise for the user. Percentage of disk and CPU performance, idle detection, power status detection etc I don't see happening.

**Comment by scienmind** [comment]
@Avamander atm you can use standard OS tools (i.e task manager -> processes) to lower the syncthing processes' priority and/or affinity (limit it to use only subset of CPU cores), also lowering I/O priority helps a lot (though not all standard OS tools allow that). I use this to make "weak" machines accessible/usable even during heavy syncthing's operations.

P.S. If you're on windows, I would also recommend "[Process Lasso](https://bitsum.com/)" for such purposes, it allows automation of such stuff, and lots of fine-tunings (manage I/O priority, etc.). It's not FOSS, and some extras are paid, but the SW is very nicely executed (the only "optimization SW" i know that actually does stuff properly) and run by single developer, so have a look.

**Comment by scienmind** [comment]
BTW, thinking of that, maybe the basic "process base priority" "process I/O priority" "process affinity" settings (or subset of it) may be incorporated into syncthing settings itself without much trouble. Some wrappers already doing that. 
Though idk, maybe it'd require too much OS-specific code for each OS and would make no sense to include it in core ST...

**Comment by calmh** [comment]
If there are reasonably simple APIs to say the equivalent of "I am a background process" and thus get worse scheduling priority etc I'm all for doing that in all cases, no user tweaks required.

**Comment by scienmind** [comment]
Seem to be reasonably simple:
Info for windows: [MSDN](https://msdn.microsoft.com/en-us/library/ms686219%28v=vs.85%29.aspx)
Linux: has [nice](http://man7.org/linux/man-pages/man1/nice.1.html) / [ionice](http://man7.org/linux/man-pages/man1/ionice.1.html) functions for setting user-level priorities (and [setpriority](http://man7.org/linux/man-pages/man2/getpriority.2.html) / [ioprio_set](http://man7.org/linux/man-pages/man2/ioprio_set.2.html) for kernel level) and affinity functions are also present.

**Comment by calmh** [comment]
Setting the process priority to "background" on Windows with that API would probably be a good default. I don't see this being as much of an issue on the other platforms most of the time.

**Comment by Avamander** [comment]
You only notice this with huge folders and slow computers. The niceness
should also be lowered when on linux by default.

**Comment by scienmind** [comment]
I don't have much ST experience on linux machines, but on windows:
If working with huge data sets (~few TB), the system slowdown is very noticeable, especially during startup, even on somewhat decent computers (i7 2nd gen CPUs).

Usually just lowering base priority is not enough, the IO traffic (huge db read & occasional scans) is chocking the whole system. So from my experience: lower IO priority + lower base priority + 50% CPU affinity makes everything smooth and nice. 

I don't think we should limit affinity by default, but [lower IO priority + lower base priority] would be a good default IMHO.

**Comment by canton7** [comment]
From MSDN: 

> processes that interact with the user should not use PROCESS_MODE_BACKGROUND_BEGIN

Give me a mention if you implement something around this: I'm currently setting Syncthing's priority when I start it, so I'll have to figure out a way to not collide horribly with what you decide to do (which may be as simple as removing my functionality).

**Comment by calmh** [comment]
What are you setting canton7? if you're already doing this it seems to me like we are covered. The thing i wrote for limiting scans could potentially be useful for limiting impact on servers, although I'm skeptical really.

**Comment by canton7** [comment]
I've got a setting to set the priority of the syncthing process, using the mechanism described above. I don't currently have anything to set IO priority or CPU affinity.

**Comment by meonkeys** [comment]
Not sure if it'll help y'all, but I [just discovered](https://docs.syncthing.net/users/faq.html#why-does-it-use-so-much-cpu) I can keep my 4-core laptop usable during initial scans by setting `GOPROCS=2` on my 4-core laptop.

Cc: #2760

**Comment by Ferroin** [comment]
It's worth pointing out that the GOMAXPROCS thing has some specific limitations:
1. It can only limit to an integer number of processes, which means you can only limit to an integral number of whole cores used at most, which means it's not particularly useful on small systems such as those used by most people according to the stats page (and yes, I do consider an 8 core machine a small system).  Even if you have 8 cores, this only lets you limit to effectively 12.5% of the system's processing power.
2. It only directly limits CPU usage, and in practice, this means it doesn't limit how fast directories are scanned.  Directory scanning will almost always have more impact on overall system performance than the hashing will, because stat() (or whatever equivalent is used on the system in question) is almost always one of the most expensive syscalls.
3. This does not affect scheduling priority at all, which means it also has limited impact on how much syncthing affects the rest of the system.  If you're running a computer game or some CAD software, or even something like a BOINC application, this won't help much, because they will still be contending with syncthing for the CPU's it is allowed to use.

**Comment by dcarosone** [comment]
I think something missed from the discussion about platform and niceness
methods for the process as a whole, was the idea of being per folder.

The issue that one (large, bulky, low priority) folder can starve out
updates in another (smaller, more active, more critical) folder won't be
helped by any external throttling of the process as a whole.

Instead of a rate limit, another way to think of this may be as relative
folder priority or weighting: something like "hash n files / blocks" from
this folder per "turn".

We already have a per-folder (iirc) number-of-hashers that will do
basically this in practice.

OP: does setting this property to 1 or 2 for the folder in question help?

**Comment by Avamander** [comment]
Does setting what property help?

**Comment by usego** [comment]
Some IO limitation per folder sounds like a good idea. I have few small but really important folders in terms of their update time like SQL replication data and also few folders with huge backups and static data (both huge in terms of file size and file count) and syncthing process startup is a pain now for smaller / important folders.

**Comment by Avamander** [comment]
I needed to sync a 11GB folder today, my PC has been unusable for like 6h,
because it started scanning AGAIN after it finished. There has to be a
better way to monitor for changes and sync. Dropbox synced the same folder
but it never made my PC unusable.

**Comment by AudriusButkevicius** [comment]
Check out syncthing-inotify.

I am locking this, as latest comments are either rants or of a +1 nature which do not really contribute to the issue.

**Comment by AudriusButkevicius** [comment]
We now have an option to specify IO scheduling class, which means the OS will not grant syncthing IO unless there is nothing else running, which addresses the issue OP was observing.

Furthermore, we now have options tweak concurrent scans, concurrent syncs etc, which should provide sufficient control on top of that.

I don't think we'll ever add support for "scan if disk utilisation is less than X%", because X is a moving target,and we'd have to stop and start activity abruptly. Furthermore, we  ourselves affect X, causing our own actions to stop ourselves.

I've unlocked the issue in case anyone has further comments, but I think existing controls are sufficient, hence closing this.
